### PR TITLE
[move][sui-adapter][sui-framework] Remove object versioning from the Move layer

### PR DESCRIPTION
**Pull request by tnowacki**
- Removed version from object::Info
- Renamed object::Info to object::UID
- The version of objects is now tracked on the Rust side in MoveObject

Blocked by #3618

More background:

- We were going to add a `child_count` alongside the version in `VersionedID` in Move, this motivated the `object::Info` renaming
- After attempting to do `child_count`, we realized that neither version nor child count can  be exposed in Move due to cases where they are inaccurate during the transaction. So we are moving them into the Rust layer in order to reduce the gas cost of Move transactions and to reduce the complexity for Move devs
- However, when unwrapping wrapped objects, to either delete them or transfer them, you need to know the previous version. Before removing the `version` from Move, we can essentially just double increment the version. But after it is moved into the Rust layer, we don’t know the version in the Move data after unwrapping. So we will grab the last version from the `parent_sync` for these unwrapped objects.

### CODE REVIEW

**Comment by tnowacki** [comment]
@tharbert This PR (I believe) requires regenerating genesis, how do I do that?

@huitseeker @mwtian I figured out this would break this with `crates/sui-config/tests/snapshot_tests.rs`, which was nice :). But it only tells me how to regenerate the test snapshot, not the underlying config/genesis blob.

File: crates/sui-framework/sources/object.move [hunk 1]
```diff
@@ -4,8 +4,6 @@
 /// Sui object identifiers
```
Comment (tnowacki) [inline crates/sui-framework/sources/object.move:4]: This file contains the important Move changes. Everything else is just shifting to fit the API changes here.

File: crates/sui-types/src/object.rs [hunk 2]
```diff
@@ -72,6 +69,7 @@ impl MoveObject {
     pub unsafe fn new_from_execution(
         type_: StructTag,
         has_public_transfer: bool,
+        version: SequenceNumber,
```
Comment (tnowacki) [inline crates/sui-types/src/object.rs:72]: This is probably the biggest change for MoveObject. Lots of downstream changes as a result of this

File: crates/sui-adapter/src/adapter.rs [hunk 3]
```diff
@@ -562,95 +582,112 @@ fn handle_transfer<
...
-            let has_public_transfer = abilities.has_store();
-            // safe because `has_public_transfer` was properly determined from the abilities
-            let mut move_obj =
-                unsafe { MoveObject::new_from_execution(s_type, has_public_transfer, contents) };
-            let old_object = by_value_objects.remove(&move_obj.id());
-
-            #[cfg(debug_assertions)]
-            {
-                check_transferred_object_invariants(&move_obj, &old_object)
+    let s_type = match type_ {
```
Comment (tnowacki) [inline crates/sui-adapter/src/adapter.rs:585]: This isn't really that big of a change. Just some code styling changes to make it easier to read.

File: crates/sui-adapter/src/adapter.rs [hunk 4]
```diff
@@ -500,16 +509,27 @@ fn process_successful_execution<
...
+                            // we force it to `v+2` to make it sure it is distinct
+                            // This incrementing will be done in `delete_object`
                             state_view.log_event(Event::delete_object(
                                 module_id.address(),
                                 module_id.name(),
                                 ctx.sender(),
                                 *obj_id,
                             ));
+                            let previous_version =
+                                match state_view.get_latest_parent_entry_ref(*obj_id) {
```
Comment (tnowacki) [inline crates/sui-adapter/src/adapter.rs:521]: @lxfind Let me know what you think of this. Not really sure what else to do if the object isn't in parent_sync

**Review by awelc — COMMENTED** [review]
Should the docs be updated as part of the same PR to remove the discrepancy right away?

**Comment by tnowacki** [comment]
> Should the docs be updated as part of the same PR to remove the discrepancy right away?

I can if you want to. But I have been working on a separate pass through the docs.

**Comment by awelc** [comment]
> I can if you want to. But I have been working on a separate pass through the docs.

Not necessarily. I just wanted to raise awareness here so that things don't stay out-of-sync for too long.

File: crates/sui-framework/sources/object.move [hunk 5]
```diff
@@ -39,139 +29,119 @@ module sui::object {
...
-    /// - The number of child objects
+    /// Globally unique IDs that define an object's ID in storage. Any Sui Object, that is a struct
+    /// with the `key` ability, must have `id: UID` as its first field.
+    /// These are globaly unique in the sense that no two values of type `UID` are ever equal, in
+    /// other words for any two values `id1: UID` and `id2: UID`, `id1` != `id2`.
     /// This is a privileged type that can only be derived from a `TxContext`.
-    /// `Info` doesn't have the `drop` ability, so deleting a `Info` requires a call to `delete`
-    struct Info has store {
+    /// `UID` doesn't have the `drop` ability, so deleting a `UID` requires a call to `delete`.
+    struct UID has store {
```
Comment (emmazzz) [inline crates/sui-framework/sources/object.move:38]: I like the new name and the simplification here!

**Comment by mwtian** [comment]
> @tharbert This PR (I believe) requires regenerating genesis, how do I do that?
> 
> @huitseeker @mwtian I figured out this would break this with `crates/sui-config/tests/snapshot_tests.rs`, which was nice :). But it only tells me how to regenerate the test snapshot, not the underlying config/genesis blob.

I can put a link to genesis blob regeneration around the test, if it is available.

File: crates/sui-adapter/src/adapter.rs [hunk 6]
```diff
@@ -493,20 +502,20 @@ fn process_successful_execution<
...
-                let info: Info = bcs::from_bytes(&event_bytes).unwrap();
-                let obj_id = info.object_id();
+                let uid: UID = bcs::from_bytes(&event_bytes).unwrap();
+                let obj_id = uid.object_id();
                 // We don't care about IDs that are generated in this same transaction
                 // but only to be deleted.
                 if !newly_generated_ids.contains(obj_id) {
-                    match by_value_objects.remove(info.object_id()) {
-                        Some(_) => {
+                    match by_value_objects.remove(uid.object_id()) {
```
Comment (lxfind) [inline crates/sui-adapter/src/adapter.rs:510]:
```suggestion
                    match by_value_objects.remove(obj_id) {
```

File: crates/sui-adapter/src/adapter.rs (hunk 4 above)
Comment (lxfind) [reply crates/sui-adapter/src/adapter.rs:536]: Good q. We probably don't want to return `ExecutionError` here. Say it failed due to a transient db failure: we don't want to charge gas for this. Feel free to fix this in a different PR, but I think we now need to distinguish between system error and execution error in this function when return (I recall you mentioned even Move execution may return system error too, so we could fix this with that together if we want)

File: crates/sui-adapter/src/adapter.rs [hunk 7]
```diff
@@ -515,16 +524,27 @@ fn process_successful_execution<
                             // with version `v+1` along with OBJECT_DIGEST_WRAPPED. Now when the object is unwrapped,
                             // it will also have version `v+1`, leading to a violation of the invariant that any
```
Comment (lxfind) [inline crates/sui-adapter/src/adapter.rs:525]: The comments here need to be updated.

**Review by lxfind — APPROVED** [review]
Overall LGTM. Thank you so much for the heroic PR!

File: crates/sui-adapter/src/adapter.rs (hunk 4 above)
Comment (tnowacki) [reply crates/sui-adapter/src/adapter.rs:536]: I think that will be a much large fix yeah, as we don't have a mechanism for such an error right now (I think). 

I'll slap a TODO on it and open an issue

**Pull request merged**